src/hal/src/command/mod.rs

```diff
@@ -10,58 +10,123 @@ mod raw;
 mod renderpass;
 mod transfer;
 
+
 pub use self::graphics::*;
-pub use self::raw::RawCommandBuffer;
+pub use self::raw::{RawCommandBuffer, CommandBufferFlags};
 pub use self::renderpass::*;
 pub use self::transfer::*;
 
-/// Thread-safe finished command buffer for submission.
-pub struct Submit<B: Backend, C>(pub(crate) B::CommandBuffer, PhantomData<C>);
-unsafe impl<B: Backend, C> Send for Submit<B, C> {}
+/// Trait indicating how many times a Submit can be submitted.
+pub trait Shot { 
+    ///
+    fn flags() -> CommandBufferFlags;
+}
+/// Indicates a Submit that can only be submitted once.
+pub enum OneShot { }
+impl Shot for OneShot { fn flags() -> CommandBufferFlags { CommandBufferFlags::ONE_TIME_SUBMIT } }
+/// Indicates a Submit that can be submitted multiple times.
+pub enum MultiShot { }
+impl Shot for MultiShot { fn flags() -> CommandBufferFlags { Default::default() } }
+
+/// A trait indicating the level of a command buffer.
+pub trait Level { }
+/// Indicates a primary command buffer.
+pub enum Primary { }
+impl Level for Primary { }
+/// Indicates a secondary command buffer.
+pub enum Secondary { }
+impl Level for Secondary { }
 
-impl<B: Backend, C> Submit<B, C> {
+/// A trait representing a command buffer that can be added to a `Submission`.
+pub unsafe trait Submittable<B: Backend, C, L: Level> {
     ///
-    pub unsafe fn new(buffer: CommandBuffer<B, C>) -> Self {
-        Submit(buffer.raw.clone(), PhantomData)
+    unsafe fn into_buffer(self) -> B::CommandBuffer;
+}
+
+/// Thread-safe finished command buffer for submission.
+pub struct Submit<B: Backend, C, S: Shot, L: Level> {
+    pub(crate) buffer:  B::CommandBuffer,
+    pub(crate) _capability: PhantomData<C>,
+    pub(crate) _shot: PhantomData<S>,
+    pub(crate) _level: PhantomData<L>
+}
+impl<B: Backend, C, S: Shot, L: Level> Submit<B, C, S, L> {
+    fn new(buffer: B::CommandBuffer) -> Self {
+        Submit {
+            buffer: buffer,
+            _capability: PhantomData,
+            _shot: PhantomData,
+            _level: PhantomData,
+        }
     }
 }
 
+unsafe impl<B: Backend, C, L: Level> Submittable<B, C, L> for Submit<B, C, OneShot, L> {
+    unsafe fn into_buffer(self) -> B::CommandBuffer { self.buffer }
+}
+unsafe impl<'a, B: Backend, C, L: Level> Submittable<B, C, L> for &'a Submit<B, C, MultiShot, L> {
+    unsafe fn into_buffer(self) -> B::CommandBuffer { self.buffer.clone() }
+}
+unsafe impl<B: Backend, C, S: Shot, L: Level> Send for Submit<B, C, S, L> {}
```

Comment:
and impls like this are going to be simpler if the struct doesn't have as many generic bounds